### PR TITLE
Add column obfuscator for prepared statement bound values

### PR DESCRIPTION
**Pull request by etedpet**

### CODE REVIEW

File: ecaudit/src/test/java/com/ericsson/bss/cassandra/ecaudit/entry/suppressor/TestHideAllSuppressor.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+ * Tests the {@link HideAllSuppressor} class.
+ */
+@RunWith(MockitoJUnitRunner.StrictStubs.class)
+public class TestHideAllSuppressor
+{
+    @Mock
+    ByteBuffer valueMock;
+
+    @Test
+    public void testSuppressorAlwaysReturnsType()
```
Comment (emolsson) [inline ecaudit/src/test/java/com/ericsson/bss/cassandra/ecaudit/entry/suppressor/TestHideAllSuppressor.java]: Should we run parameterized tests to make sure that many types are handled in this way?

File: ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/entry/suppressor/HideBlobsSuppressor.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+               ? Optional.of("<" + column.type.asCQL3Type() + ">")
+               : Optional.empty();
+    }
+
+    private static boolean isBlobType(AbstractType<?> type)
+    {
+        return type.asCQL3Type() == CQL3Type.Native.BLOB;
+    }
+
+    private static boolean isCollectionContainingBlobType(AbstractType<?> type)
```
[the remark this answers is not in the document]
Comment (etedpet) [reply ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/entry/suppressor/HideBlobsSuppressor.java]: Refactored SuppressBlobs to look for blobs inside UDT/Tuples as well.

File: ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/entry/suppressor/HideBlobsSuppressor.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    public Optional<String> suppress(ColumnSpecification column, ByteBuffer value)
+    {
+        return isBlobType(column.type) || isCollectionContainingBlobType(column.type)
+               ? Optional.of("<" + column.type.asCQL3Type() + ">")
+               : Optional.empty();
+    }
+
+    private static boolean isBlobType(AbstractType<?> type)
+    {
+        return type.asCQL3Type() == CQL3Type.Native.BLOB;
```
[the remark this answers is not in the document]
Comment (etedpet) [reply ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/entry/suppressor/HideBlobsSuppressor.java]: removed

File: conf/audit.yaml [hunk 4]
```diff
@@ -89,3 +89,33 @@ logger_backend:
 #      - log_format: "{?client:'${CLIENT_IP}'|?}user:'${USER}'{?|batchId:'${BATCH_ID}'?}|status:'${STATUS}'|operation:'${OPERATION}'"
 #        time_format:
 #        time_zone:
+
+
+# The column suppressor determines how column values should be logged for prepared statements (i.e. bound values).
+# The suppressor is applied to all prepared statements regardless of which table is used.
+#
+# - ShowAllSuppressor           -> Will not suppress. Prints all values for prepared statements.
```
[the remark this answers is not in the document]
Comment (etedpet) [reply conf/audit.yaml]: used these names

File: conf/audit.yaml [hunk 5]
```diff
@@ -89,3 +89,33 @@ logger_backend:
...
+
+
+# The bound value suppressor determines how bound values should be logged for prepared statements.
+# The suppressor is applied to all prepared statements regardless of which table is used.
+#
+# - SuppressNothing              -> Will not suppress. Prints all values for prepared statements.
+#
+# - SuppressEverything           -> Will suppress all values by replacing the values with its type, e.g. <int>, <text>
+#
+# - SuppressorBlobs              -> Will suppress all blob values. Values will be replaced by a "<blob>"-tag in the log.
```
Comment (etedpet) [inline conf/audit.yaml]: SuppressBlobs

File: conf/audit.yaml [hunk 6]
```diff
@@ -91,31 +91,31 @@ logger_backend:
...
+# Note - The suppressor only impacts the log message if the OPERATION field is logged (see log_format above).
 # If OPERATION_NAKED is selected then no bound values will be logged.
 #
-# A custom column suppressor can be plugged in by setting the full name of an suppressor class implementing the
-# ColumnSuppressor interface.
+# A custom bound value suppressor can be plugged in by setting the full name of an suppressor class implementing the
+# BoundValueSuppressor interface.
 #
-# By default the ShowAllSuppressor will be used.
+# By default the SuppressNothing will be used.
```
Comment (emolsson) [inline conf/audit.yaml]: By default ~~the~~ SuppressNothing will be used.

File: conf/audit.yaml [hunk 7]
```diff
@@ -89,3 +89,34 @@ logger_backend:
...
+#
+# Note - The suppressor only impacts the log message if the OPERATION field is logged (see log_format above).
+# If OPERATION_NAKED is selected then no bound values will be logged.
+#
+# A custom bound value suppressor can be plugged in by setting the full name of an suppressor class implementing the
+# BoundValueSuppressor interface.
+#
+# By default the SuppressNothing will be used.
+#
+# bound_value_suppressor: SuppressorBlobs
```
Comment (emolsson) [inline conf/audit.yaml]: SuppressBlobs

File: doc/slf4j_logger.md [hunk 8]
```diff
@@ -84,6 +84,17 @@ Which will generate logs entries like this (assuming Logback pattern does not co
...
 
+### Bound Value Suppressor
+
+A bound value suppressor defining how column values should be logged for prepared statements. Using one of the non-default suppressors
+can be useful to avoid logging sensitive or less useful information, e.g. large blobs.
+
+For more information about the available suppressors, see the ```audit.yaml``` reference documentation.
+The example below will suppress/hide blobs values from being logged.
+```YAML
+column_suppressor: HideBlobsSuppressor
```
Comment (emolsson) [inline doc/slf4j_logger.md]: bound_value_suppressor: SuppressBlobs

File: ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/AuditAdapter.java [hunk 9]
```diff
@@ -261,8 +261,14 @@ public Auditor getAuditor()
...
     @VisibleForTesting
-    public void setColumnSuppressor(ColumnSuppressor columnSuppressor)
+    public void setBoundValueSuppressor(BoundValueSuppressor boundValueSuppressor)
     {
-        this.columnSuppressor = columnSuppressor;
+        this.boundValueSuppressor = boundValueSuppressor;
+    }
+
+    @VisibleForTesting
+    public BoundValueSuppressor  getBoundValueSuppressor()
```
Comment (emolsson) [inline ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/AuditAdapter.java]: Nit: double spaces

File: ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/entry/suppressor/SuppressBlobs.java [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+        switch(type.kind)
+        {
+            case LIST:
+                return listContainsBlob((ListType) type);
+            case SET:
+                return setContainsBlob((SetType) type);
+            case MAP:
+                return mapContainsBlob((MapType) type);
+            default:
+                throw new IllegalArgumentException("Invalid collection type");
```
Comment (emolsson) [inline ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/entry/suppressor/SuppressBlobs.java]: Include type.kind in the error message?

File: ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/entry/suppressor/SuppressEverything.java [hunk 11]
```diff
@@ -20,11 +20,20 @@
...
     {
-        return Optional.of("<" + column.type.asCQL3Type() + ">"); // All values should be suppressed
+        return Optional.of(suppressWithType(column)); // All values should be suppressed
+    }
+
+    /**
+     * @param column the column specification
+     * @return the string representation of the suppressed column type
+     */
+    static final String suppressWithType(ColumnSpecification column)
```
Comment (emolsson) [inline ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/entry/suppressor/SuppressEverything.java]: Nit: would it make sense to have this elsewhere? All other suppressors seems to be dependent on it.
I.e. we could change the interface to be an abstract class instead.

I am fine with keeping it here as well, just a thought though.

File: conf/audit.yaml (hunk 6 above)
Comment (etedpet) [reply conf/audit.yaml]: done

File: conf/audit.yaml (hunk 7 above)
Comment (etedpet) [reply conf/audit.yaml]: done

File: doc/slf4j_logger.md (hunk 8 above)
Comment (etedpet) [reply doc/slf4j_logger.md]: done

File: ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/AuditAdapter.java (hunk 9 above)
Comment (etedpet) [reply ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/AuditAdapter.java]: done

File: ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/entry/suppressor/SuppressBlobs.java (hunk 10 above)
Comment (etedpet) [reply ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/entry/suppressor/SuppressBlobs.java]: done

File: ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/entry/suppressor/SuppressEverything.java (hunk 11 above)
Comment (etedpet) [reply ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/entry/suppressor/SuppressEverything.java]: I'll keep the interface but create an abstract class for these suppress-classes...

File: ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/entry/suppressor/AbstractSuppressor.java [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+
+import org.apache.cassandra.cql3.ColumnSpecification;
+
+public abstract class AbstractSuppressor implements BoundValueSuppressor
+{
+    /**
+     * @param column the column specification
+     * @return the string representation of the suppressed column type
+     */
+    String suppressWithType(ColumnSpecification column)
```
Comment (emolsson) [inline ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/entry/suppressor/AbstractSuppressor.java]: protected? (Available to all subclasses, not just our own)

Comment (etedpet) [reply ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/entry/suppressor/AbstractSuppressor.java]: ok

**Pull request merged**